tests/checkers/unittest_unicode/bad_chars.py

```diff
@@ -0,0 +1,255 @@
+                end_line=2 + start_lines,
+                # node=module,
+                args=None,
+                confidence=pylint.interfaces.HIGH,
+                col_offset=27,
+                end_col_offset=28,
+            ),
+            pylint.testutils.MessageTest(
+                msg_id="invalid-character-carriage-return",
+                line=3 + start_lines,
+                end_line=3 + start_lines,
+                # node=module,
+                args=None,
+                confidence=pylint.interfaces.HIGH,
+                col_offset=23,
+                end_col_offset=24,
+            ),
+            pylint.testutils.MessageTest(
+                msg_id="invalid-character-sub",
+                line=4 + start_lines,
+                end_line=4 + start_lines,
+                # node=module,
+                args=None,
+                confidence=pylint.interfaces.HIGH,
+                col_offset=21,
+                end_col_offset=22,
+            ),
+            pylint.testutils.MessageTest(
+                msg_id="invalid-character-esc",
+                line=5 + start_lines,
+                end_line=5 + start_lines,
+                # node=module,
+                args=None,
+                confidence=pylint.interfaces.HIGH,
+                col_offset=21,
+                end_col_offset=22,
+            ),
+        ]
+        with self.assertAddsMessages(*expected):
+            self.checker.process_module(module)
+
+    @pytest.mark.parametrize(
+        "codec_and_msg, char, msg_id",
+        [
+            pytest.param(
+                codec_and_msg,
+                char_msg[0],
+                char_msg[1],
+                id=f"{char_msg[1]}_{codec_and_msg[0]}",
+            )
+            for codec_and_msg, char_msg in itertools.product(
+                CODEC_AND_MSG,
+                (
+                    ("\0", "invalid-character-nul"),
+                    ("\N{ZERO WIDTH SPACE}", "invalid-character-zero-width-space"),
+                ),
+            )
+            # Only utf contains zero width space
+            if (
+                char_msg[0] != "\N{ZERO WIDTH SPACE}"
+                or codec_and_msg[0].startswith("utf")
+            )
+        ],
+    )
+    def test_bad_chars_that_would_currently_crash_python(
+        self,
+        char: str,
+        msg_id: str,
+        tmp_path: Path,
+        codec_and_msg: Tuple[str, Tuple[pylint.testutils.MessageTest]],
+    ):
+        """Special test for a file containing chars that lead to
+
+        Python or Astroid crashes (which causes Pylint to exit early)
+        """
+        codec, start_msg = codec_and_msg
+        test_file = tmp_path / "failing.py"
+        test_file.write_text(
+            f"# # coding: {codec}\n"
+            f"# file containing {char} will fail loading in AST checker \n"
+            f"# but we still want to check it, in case this behavior changes \n",
+            encoding=codec,
+        )
```

Comment:
Why are we writing a comment in a temporary test file ? 

```suggestion
        test_file = tmp_path / "failing.py"
        # file containing {char} will fail loading in AST checker
        # but we still want to check it, in case this behavior changes
        test_file.write_text(f"# # coding: {codec}\n", encoding=codec)
```

Reply:
The comment needs to contain the char and needs to be valid python file.

Reply:
```suggestion
        test_file = tmp_path / "failing.py"
        test_file.write_text(
            f"""# # coding: {codec}
# file containing {char} will fail loading in AST checker
# but we still want to check it, in case this behavior changes
""",  encoding=codec
        )
```